orttraining/orttraining/test/graph/gradient_graph_builder_test.cc

```diff
@@ -798,6 +800,156 @@ class PipelineSplitter {
   }
 };
 
+// TODO: move to a proper location for pipeline training
+struct PipelineBatchInfo {
+  std::vector<std::pair<int64_t, int64_t>> events;
+  int64_t retired_batch;  // index of retired batch, so its data could be reused
+};
+
+class PipelineBatchPlanner {
+ private:
+  int64_t max_id_;
+
+ public:
+  PipelineBatchPlanner()
+      : max_id_(::onnxruntime::contrib::OrtEventPool::GetPoolSize() - 1) {
+  }
+
+  void CreatePlan(const std::vector<int64_t>& start_ids,
```

Comment:
>CreatePlan [](start = 7, length = 10)

In the real world case, each stage is ran with one GPU and handled by one process. 

The plan for only one stage is needed in a process. In such case, the planning process would be very simple for each stage. 

After the global timeline is generated, a process scans its own stage from left to right and generate something like 
[-1, 0] -> [0, 1] -> [1, 2] -> [2, 3] for the work items in this stage.